File: financial-statement-studio/apps/api/app/schemas/notes.py
from datetime import datetime
from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


NoteType = Literal[
    "accounting_policy",
    "statement_note",
    "general_disclosure",
]

StatementName = Literal[
    "profit_or_loss",
    "financial_position",
    "cash_flows",
    "changes_in_equity",
]


class DisclosureTemplateCreate(BaseModel):
    template_key: str = Field(
        min_length=1,
        max_length=120,
    )

    title: str = Field(
        min_length=1,
        max_length=255,
    )

    note_type: NoteType = (
        "general_disclosure"
    )

    statement_name: (
        StatementName | None
    ) = None

    statement_line_key: str | None = (
        Field(
            default=None,
            max_length=100,
        )
    )

    default_content: str = ""

    is_required: bool = False
    is_active: bool = True

    display_order: int = Field(
        default=0,
        ge=0,
    )


class DisclosureTemplateUpdate(BaseModel):
    template_key: str | None = Field(
        default=None,
        min_length=1,
        max_length=120,
    )

    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
    )

    note_type: NoteType | None = None

    statement_name: (
        StatementName | None
    ) = None

    statement_line_key: str | None = (
        Field(
            default=None,
            max_length=100,
        )
    )

    default_content: str | None = None

    is_required: bool | None = None
    is_active: bool | None = None

    display_order: int | None = Field(
        default=None,
        ge=0,
    )


class DisclosureTemplateResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: str

    template_key: str
    title: str
    note_type: str

    statement_name: str | None
    statement_line_key: str | None

    default_content: str

    is_system_template: bool
    is_required: bool
    is_active: bool

    display_order: int

    created_at: datetime
    updated_at: datetime


class DisclosureTemplateListResponse(
    BaseModel,
):
    items: list[
        DisclosureTemplateResponse
    ]

    total: int


class DisclosureTemplateInitializationResponse(
    BaseModel,
):
    created_count: int
    skipped_count: int

    items: list[
        DisclosureTemplateResponse
    ]


class FinancialReportNoteCreate(BaseModel):
    template_id: str | None = None

    note_number: int | None = Field(
        default=None,
        ge=1,
    )

    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
    )

    note_type: NoteType | None = None

    statement_name: (
        StatementName | None
    ) = None

    statement_line_key: str | None = (
        Field(
            default=None,
            max_length=100,
        )
    )

    content: str | None = None

    is_active: bool = True


class FinancialReportNoteUpdate(BaseModel):
    template_id: str | None = None

    note_number: int | None = Field(
        default=None,
        ge=1,
    )

    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
    )

    note_type: NoteType | None = None

    statement_name: (
        StatementName | None
    ) = None

    statement_line_key: str | None = (
        Field(
            default=None,
            max_length=100,
        )
    )

    content: str | None = None

    is_active: bool | None = None


class FinancialReportNoteResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: str
    financial_report_id: str
    template_id: str | None

    note_number: int

    title: str
    note_type: str

    statement_name: str | None
    statement_line_key: str | None

    content: str

    is_active: bool

    created_at: datetime
    updated_at: datetime


class FinancialReportNoteListResponse(
    BaseModel,
):
    financial_report_id: str

    items: list[
        FinancialReportNoteResponse
    ]

    total: int


class ReportNotesInitializationRequest(
    BaseModel,
):
    include_optional: bool = True


class ReportNotesInitializationResponse(
    BaseModel,
):
    financial_report_id: str

    created_count: int
    skipped_count: int

    items: list[
        FinancialReportNoteResponse
    ]


class ReorderFinancialReportNotesRequest(
    BaseModel,
):
    note_ids: list[str] = Field(
        min_length=1,
    )

File: financial-statement-studio/apps/web/src/types/notes.ts
export const NOTE_TYPE_OPTIONS = [
  {
    value: "accounting_policy",
    label: "Accounting policy",
  },
  {
    value: "statement_note",
    label: "Statement note",
  },
  {
    value: "general_disclosure",
    label: "General disclosure",
  },
] as const;

export const STATEMENT_NAME_OPTIONS = [
  {
    value: "profit_or_loss",
    label: "Profit or Loss",
  },
  {
    value: "financial_position",
    label: "Financial Position",
  },
  {
    value: "cash_flows",
    label: "Cash Flows",
  },
  {
    value: "changes_in_equity",
    label: "Changes in Equity",
  },
] as const;

export type NoteType =
  (typeof NOTE_TYPE_OPTIONS)[number]["value"];

export type StatementName =
  (typeof STATEMENT_NAME_OPTIONS)[number]["value"];

export const NOTE_TYPE_LABELS: Record<
  NoteType,
  string
> = {
  accounting_policy: "Accounting policy",
  statement_note: "Statement note",
  general_disclosure: "General disclosure",
};

export const STATEMENT_NAME_LABELS: Record<
  StatementName,
  string
> = {
  profit_or_loss: "Profit or Loss",
  financial_position: "Financial Position",
  cash_flows: "Cash Flows",
  changes_in_equity: "Changes in Equity",
};

export type DisclosureTemplate = {
  id: string;

  template_key: string;
  title: string;
  note_type: NoteType;

  statement_name: StatementName | null;
  statement_line_key: string | null;

  default_content: string;

  is_system_template: boolean;
  is_required: boolean;
  is_active: boolean;

  display_order: number;

  created_at: string;
  updated_at: string;
};

export type DisclosureTemplateListResponse = {
  items: DisclosureTemplate[];
  total: number;
};

export type DisclosureTemplateInitializationResponse = {
  created_count: number;
  skipped_count: number;
  items: DisclosureTemplate[];
};

export type DisclosureTemplateCreatePayload = {
  template_key: string;
  title: string;
  note_type: NoteType;

  statement_name?: StatementName | null;
  statement_line_key?: string | null;

  default_content?: string;

  is_required?: boolean;
  is_active?: boolean;

  display_order?: number;
};

export type DisclosureTemplateUpdatePayload =
  Partial<DisclosureTemplateCreatePayload>;

export type FinancialReportNote = {
  id: string;
  financial_report_id: string;
  template_id: string | null;

  note_number: number;

  title: string;
  note_type: NoteType;

  statement_name: StatementName | null;
  statement_line_key: string | null;

  content: string;

  is_active: boolean;

  created_at: string;
  updated_at: string;
};

export type FinancialReportNoteListResponse = {
  financial_report_id: string;
  items: FinancialReportNote[];
  total: number;
};

export type ReportNotesInitializationResponse = {
  financial_report_id: string;
  created_count: number;
  skipped_count: number;
  items: FinancialReportNote[];
};

export type FinancialReportNoteCreatePayload = {
  template_id?: string | null;
  note_number?: number;

  title?: string;
  note_type?: NoteType;

  statement_name?: StatementName | null;
  statement_line_key?: string | null;

  content?: string;

  is_active?: boolean;
};

export type FinancialReportNoteUpdatePayload =
  Partial<FinancialReportNoteCreatePayload>;

export type ReorderFinancialReportNotesPayload = {
  note_ids: string[];
};